SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,278 @@
+	uint32 MinimumRequiredWorkers = 0;
+	for (const auto& Elem : LayerNameToLBStrategy)
+	{
+		MinimumRequiredWorkers += Elem.Value->GetMinimumRequiredWorkers();
+	}
+
+	UE_LOG(LogLayeredLBStrategy, Verbose, TEXT("LayeredLBStrategy needs %d workers to support all layer strategies."), MinimumRequiredWorkers);
+	return MinimumRequiredWorkers;
+}
+
+void ULayeredLBStrategy::SetVirtualWorkerIds(const VirtualWorkerId& FirstVirtualWorkerId, const VirtualWorkerId& LastVirtualWorkerId)
+{
+	// If the LayeredLBStrategy wraps { SingletonStrategy, 2x2 grid, Singleton } and is given IDs 1 through 6 it will assign:
+	// Singleton : 1
+	// Grid : 2 - 5
+	// Singleton: 6
+	VirtualWorkerId NextWorkerIdToAssign = FirstVirtualWorkerId;
+	for (auto& Elem : LayerNameToLBStrategy)
+	{
+		UAbstractLBStrategy* LBStrategy = Elem.Value;
+		VirtualWorkerId MinimumRequiredWorkers = LBStrategy->GetMinimumRequiredWorkers();
+
+		VirtualWorkerId LastVirtualWorkerIdToAssign = NextWorkerIdToAssign + MinimumRequiredWorkers - 1;
+		if (LastVirtualWorkerIdToAssign > LastVirtualWorkerId)
+		{
+			UE_LOG(LogLayeredLBStrategy, Error, TEXT("LayeredLBStrategy was not given enough VirtualWorkerIds to meet the demands of the layer strategies."));
+			return;
+		}
+		UE_LOG(LogLayeredLBStrategy, Log, TEXT("LayeredLBStrategy assigning VirtualWorkerIds %d to %d to Layer %s"), NextWorkerIdToAssign, LastVirtualWorkerIdToAssign, *Elem.Key.ToString());
+		LBStrategy->SetVirtualWorkerIds(NextWorkerIdToAssign, LastVirtualWorkerIdToAssign);
+
+		for (VirtualWorkerId id = NextWorkerIdToAssign; id <= LastVirtualWorkerIdToAssign; id++)
+		{
+			VirtualWorkerIdToLayerName.Add(id, Elem.Key);
+		}
+
+		NextWorkerIdToAssign += MinimumRequiredWorkers;
+	}
+}
+
+FName ULayeredLBStrategy::GetLayerNameForClass(const TSubclassOf<AActor> Class) const
+{
+	if (Class == nullptr)
+	{
+		return NAME_None;
+	}
+
+	UClass* FoundClass = Class;
+	TSoftClassPtr<AActor> ClassPtr = TSoftClassPtr<AActor>(FoundClass);
+
+	while (FoundClass != nullptr && FoundClass->IsChildOf(AActor::StaticClass()))
+	{
+		if (const FName* Layer = ClassPathToLayer.Find(ClassPtr))
+		{
+			FName LayerHolder = *Layer;
+			if (FoundClass != Class)
+			{
+				ClassPathToLayer.Add(TSoftClassPtr<AActor>(Class), LayerHolder);
+			}
+			return LayerHolder;
+		}
+
+		FoundClass = FoundClass->GetSuperClass();
+		ClassPtr = TSoftClassPtr<AActor>(FoundClass);
+	}
+
+	// No mapping found so set and return default actor group.
+	ClassPathToLayer.Add(TSoftClassPtr<AActor>(Class), SpatialConstants::DefaultLayer);
+	return SpatialConstants::DefaultLayer;
+}
+
+bool ULayeredLBStrategy::IsSameWorkerType(const AActor* ActorA, const AActor* ActorB) const
+{
+	if (ActorA == nullptr || ActorB == nullptr)
+	{
+		return false;
+	}
+
+	const FName& LayerA = GetLayerNameForClass(ActorA->GetClass());
+	const FName& LayerB = GetLayerNameForClass(ActorB->GetClass());
```

Comment:
Stored as const ref here but the method returns a value?

Reply:
I could just have the return directly compare the values? Which do you think is more readable?

Reply:
Yeah, I decided to do that.